### ISSUE TITLE
Searching using extended mode doesn't handle \0 properly

### ISSUE THREAD

**Issue by aussiechip** [notepad-plus-plus/notepad-plus-plus#8270]
### Description of the Issue

In a binary file that contains a NUL, searching for \0 on its own works properly, and searches with anything before a \0 seems to work, but searching for \0 followed by anything else fails. It generally finds a \0 followed by some other character. 

### Steps to Reproduce the Issue

1. Open a file with lots of NUL characters in it. Perhaps you could create one that has \0 followed by every other ASCII character, as a test?
2. Look for a pattern that involves a \0 followed by another character.
3. If it finds the right one, try searching for a \0 followed by some different character

### Expected Behavior

I expected it to find the string I was searching for! (I happened to be looking for \0A\0P)

### Actual Behavior

It found something else

### Debug Information

Notepad++ v7.8.6   (64-bit)
Build time : Apr 21 2020 - 15:23:03
Path : C:\Program Files\Notepad++\notepad++.exe
Admin mode : OFF
Local Conf mode : OFF
OS Name : Windows 10 Pro (64-bit) 
OS Version : 1903
OS Build : 18362.418
Plugins : JSMinNPP.dll mimeTools.dll NppConverter.dll NppExport.dll

**Comment by sasumner** [comment]
> looking for \0A\0P

A potential workaround while this is being looked into is to do the search in _Regular expression_ mode.  That mode does not really like dealing with NUL characters either (and of course Notepad++ as a whole doesn't like them--and probably never will), but if your files use the most basic of encodings, you might have luck searching for `[^\x01-\xff]A[^\x01-\xff]P`.  A bit painful, but it seems to work.

**Comment by sasumner** [comment]
Actually a simpler regex search of `\x00P` should work as well.
Apparently I was thinking that regular expression searches had trouble with NUL characters as well.

**Comment by aussiechip** [comment]
OK. I had kindof assumed that that was what was happening behind the scenes anyway - the extended mode string would be internally expanded to a regexp, and then passed to the search engine. It was a pretty rash assumption, I'l admit.

As I was testing this, it became pretty clear that the search was locating random characters that happened to be in the search string memory before it was copied out of the dialog box's field. The copy of the search string from the dialog presumably terminates with the first NUL it finds, and then the search string length is passed separately, or something like that.

Anyway, that might give a bug hunter a starting point.

**Comment by aussiechip** [comment]
oops. Sorry.

**Comment by sasumner** [comment]
> the extended mode string would be internally expanded to a regexp, and then passed to the search engine

This isn't what happens.  What happens is simpler:  The extended mode string is examined for any of the special escape sequences that extended mode supports, and these sequences are replaced by actual values, e.g. `\t` is replaced by 9, `\r` by 13, `\n` by 10, etc.  `\0` is replaced by, you guessed it, zero.  

The search is then conducted in non-regex mode.

If certain routines in the call chain are set to look for a "null terminated C string" (and they are in this case), the effective end of the string is right there at the null.  Since a string and a length are both passed in to the top-level routine, part of the call chain deals with it correctly; that's why you see that a `\0` itself (or leading characters followed by a `\0`) can be matched.

The nitty-gritty detail of this problem is in the call to the Windows function `WideCharToMultiByte` with a `cchWideChar` parameter of -1.  The -1 indicates to the function that the incoming string is null-terminated. so go ahead and count on that fact.  It does its work on that basis, and ends consideration at the first `\0`, not thinking about data beyond that point.

**Comment by aussiechip** [comment]
Yeah that's pretty much what I assumed was going on, without a crystal ball and days to get used to the code structure, and being hell rusty at coding anyway. Fixing it is well beyond me, in any reasonable timeframe. Anyway, I'll let you address the most sensible fix if you feel so inclined. Seems to me that the regexp route might be the easiest. The fact that "\0" is offered as an extended search term implies that it should be handled correctly.
